WIP: AArch64: laddEvaluator

aarongraham9 opened the pull request:
WIP on AArch64 laddEvaluator.

Signed-off-by: Aaron Graham <aaron.graham@unb.ca>

aarongraham9 commented:
Related Tril Regression Tests for ladd/lsubEvaluators can be found in PR: #3214.

aarongraham9 commented:
This is good for you to take a look at @0xdaryl, @knn-k, @marufunb and @jeanlego.

0xdaryl commented on compiler/aarch64/codegen/BinaryEvaluator.cpp:
```diff
@@ -64,6 +64,42 @@ static TR::Register *addOrSubInteger(TR::Node *node, TR::CodeGenerator *cg)
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Register *trgReg = cg->allocateRegister();
+   bool isAdd = node->getOpCode().isAdd();
+
```
Some work can be done here to tighten the use of virtual registers, thereby improving the resulting register assignment.  In particular, you shouldn't be unconditionally creating a virtual for the target register if you can re-use one of the source registers whose live range ends at this instruction.

You need to check the reference count on each source node and if the reference count goes to zero then you can conclude that the live range ends here, and the register produced from evaluating that source node is eligible for re-use as the target register.  If both source registers die at this instruction then you need to make a heuristic choice on which one to re-use (likely just pick one).

aarongraham9 replied:
Ok. I'll check the reference count on the node to know when I can re-use a register.

knn-k commented on compiler/aarch64/codegen/BinaryEvaluator.cpp:
```diff
@@ -64,6 +64,42 @@ static TR::Register *addOrSubInteger(TR::Node *node, TR::CodeGenerator *cg)
    }
 
+static TR::Register *
+addOrSubLongInteger(TR::Node *node, TR::CodeGenerator *cg)
```
You can see `commonIntegerAbsEvaluator()` in `UnaryEvaluator.cpp` (already merged) and  `irolEvaluator()` in #3208 (not merged yet) as examples of commoning functions for int and long.

aarongraham9 commented:
Rebased off of `master` and commoned up the functionality in my `laddEvaluator`/`lsubEvaluator` helper function with the `iaddEvaluator`/`isubEvaluator` helper function.

aarongraham9 commented:
Adding WIP until I'm finished with analyzing/optimizing/tightening-up the use of Virtual Registers by using the reference counts.

knn-k commented:
Please see my change in #3406.

aarongraham9 commented:
Rebsed off of the `master` branch picking up the fix in #3406 from @knn-k.

All tril tests are passing for iadd, isub, ladd and lsub:
```
./comptest
[==========] Running 2056 tests from 2 test cases.
[----------] 900 tests from ArithmeticTest/Int32Arithmetic
[----------] 900 tests from ArithmeticTest/Int32Arithmetic (15459 ms total)

[----------] 1156 tests from ArithmeticTest/Int64Arithmetic
[----------] 1156 tests from ArithmeticTest/Int64Arithmetic (19553 ms total)

[==========] 2056 tests from 2 test cases ran. (35023 ms total)
[  PASSED  ] 2056 tests.
[  ALL TESTS PASSED  ]
```

aarongraham9 commented:
Testing with filters.
```
./comptest --gtest_filter=*ArithmeticTest*Int*Arithmetic*

Note: Google Test filter = *ArithmeticTest*Int*Arithmetic*
[==========] Running 2506 tests from 2 test cases.
[----------] 1350 tests from ArithmeticTest/Int32Arithmetic
[----------] 1350 tests from ArithmeticTest/Int32Arithmetic (23111 ms total)

[----------] 1156 tests from ArithmeticTest/Int64Arithmetic
[----------] 1156 tests from ArithmeticTest/Int64Arithmetic (19388 ms total)

[==========] 2506 tests from 2 test cases ran. (42527 ms total)
[  PASSED  ] 2506 tests.
[  ALL TESTS PASSED  ] 
```

Not just need to tightening-up the use of virtual registers.

aarongraham9 commented:
- Commoned-up the functionality in my `laddEvaluator` and `lsubEvaluator` Evaluators with the `iaddEvaluator` and `isubEvaluator` Evaluators.
- Tightened up the use of virtual registers. It now re-uses one if possible; otherwise it allocates a new one.  
- Rebased this PR off of the `master` branch.
- Incorporated @knn-k's fix in #3406 from the `master` branch.
- Related Tril Regression Tests for `ladd`/`lsub` OMR IL OpCodes can be found in PR: #3214.

- Tril Test Results for `iadd`, `isub`, `ladd`, and `lsub` OMR IL OpCodes/Evaluators are as follows:
`./comptest --gtest_filter=*ArithmeticTest*Int*Arithmetic*`:
```
Note: Google Test filter = *ArithmeticTest*Int*Arithmetic*icTest*Int*Arithmetic*
[==========] Running 2056 tests from 2 test cases.
[----------] 900 tests from ArithmeticTest/Int32Arithmetic
[----------] 900 tests from ArithmeticTest/Int32Arithmetic (15724 ms total)

[----------] 1156 tests from ArithmeticTest/Int64Arithmetic
[----------] 1156 tests from ArithmeticTest/Int64Arithmetic (19560 ms total)

[==========] 2056 tests from 2 test cases ran. (35313 ms total)
[  PASSED  ] 2056 tests.
```

Removing the WIP as this should be good to go unless anyone has any further comments or concerns: @0xdaryl, @knn-k, @jeanlego, @marufunb and @hsoontie.

aarongraham9 commented:
All CI builds have passed; any further comments or concerns @0xdaryl, @knn-k, @jeanlego, @marufunb and @hsoontie?

0xdaryl commented on compiler/aarch64/codegen/BinaryEvaluator.cpp:
```diff
@@ -30,33 +30,55 @@
    TR::Node *firstChild = node->getFirstChild();
-   TR::Register *src1Reg = cg->evaluate(firstChild);
+   /* Get a clobberable register (re-uses one if possible otherwise allocates one) */
+   TR::Register *src1Reg = cg->gprClobberEvaluate(firstChild);
```
A clobber evaluation for both of these source registers is not what you want here.  A clobber evaluation is useful when the register itself will be modified by the instruction but yet it's original value is still needed (because of commoning, for example).  Clobber evaluation will insert an instruction to copy the register to a temporary register if its reference count is > 1 (i.e., its value is still consumed elsewhere) and skip the copy if this is the last reference it.

Simple calls to `evaluate` are what you want here.

However, there is an opportunity to optimize the target register though.  If one of the source registers "dies" at this instruction then you can re-use it for the target register rather than allocating a new register.  So, if the reference count on either source register == 1 then you can conclude that this is the last reference to that node and its register live range ends here and can be reused.

aarongraham9 replied:
Ok fair enough. Agreed. I'll check the reference count at this point to see about re-use.

aarongraham9 commented:
- Rebased this PR off of the `master` branch.
- Updated based on review comments.
- Waiting on Travis CI and  `@genie-omr build aarch64` Jenkins CI builds.

0xdaryl commented on compiler/aarch64/codegen/BinaryEvaluator.cpp:
```diff
@@ -30,33 +30,67 @@
+
+   if (1 == firstChild->getReferenceCount())
+      {
+      trgReg = cg->evaluate(firstChild);
```
You could just use `src1Reg` here rather than calling the `evaluate` function again.

aarongraham9 replied:
Ok. I'll fix it up.

aarongraham9 commented:
Yes, once #3517 is merged that will subsume this PR. Should I try and get it in beforehand or is this PR now irrelevant @0xdaryl?

0xdaryl commented:
I don't think there's much value in merging this PR if #3517 subsumes it.  However, the lessons learned while implementing this evaluator should be applied to the more general binary evaluator.

aarongraham9 commented:
Subsumed by #3517. Closing.
